File: worker/src/storyline.ts
/** Layer 1 — ABC storyline (expectation). Port of slk_bot/slk/storyline.py.
 *  Snapshots are built point-in-time: each H4 close gets a storyline made
 *  only from data closed at that moment. */
import * as F from "./features";
import type { Candle, Direction, Storyline } from "./types";
import type { StrategyConfig } from "./config";

export const MAP_TF_SECONDS = 14400; // H4

export function buildStoryline(d1: Candle[], h4: Candle[], cfg: StrategyConfig): Storyline {
  const asof = h4.length ? h4[h4.length - 1].t : Date.now();
  const base: Storyline = {
    asof, valid: false, reason: "", direction: null,
    environment: "consolidation", phase: "range", htfAlignment: "",
    origin: null, drawOnLiquidity: null, nearestExternalTarget: null,
    internalPools: [], externalPools: [], imbalances: [], mapClose: 0,
  };
  if (h4.length < 20) return { ...base, reason: "insufficient H4 data" };
  if (d1.length < 10) return { ...base, reason: "insufficient daily data" };

  const env = F.environment(h4, cfg.pivotLeft, cfg.pivotRight, cfg.minSwingsEnv);
  const ph = F.phase(h4, env, cfg.phaseLookback, cfg.pivotLeft, cfg.pivotRight);
  const align = F.htfAlignment(d1, env, cfg.pivotLeft, cfg.pivotRight);
  const close = h4[h4.length - 1].c;

  let direction: Direction | null = null;
  if (env === "bullish") direction = "LONG";
  else if (env === "bearish") direction = "SHORT";
  const story: Storyline = { ...base, environment: env, phase: ph, htfAlignment: align, mapClose: close, direction };

  if (!direction) return { ...story, reason: "no directional H4 environment (consolidation)" };

  // HTF invalidation: close through opposing structure kills the story
  const inv = F.structureInvalidationLevel(h4, direction, cfg.pivotLeft, cfg.pivotRight);
  if (inv !== null) {
    if (direction === "SHORT" && close > inv)
      return { ...story, reason: `H4 close ${close} broke opposing structure ${inv}` };
    if (direction === "LONG" && close < inv)
      return { ...story, reason: `H4 close ${close} broke opposing structure ${inv}` };
  }

  const atrMap = F.atr(h4, cfg.atrPeriod);
  const ext = F.externalPools(d1, cfg.pivotLeft, cfg.pivotRight);
  const intp = F.internalPools(h4, atrMap, cfg.decisionAtrMult, cfg.pivotLeft, cfg.pivotRight);
  const imb = F.fvgZones(h4, cfg.fvgLookback);
  const levels = F.keyLevels(h4, cfg);
  F.markFvgOverlap(levels, imb);

  const origin = F.selectOrigin(levels, close, atrMap, direction, cfg);
  if (!origin) return { ...story, reason: "no origin key level within reach of price" };

  // draw on liquidity: nearest external pool in the storyline direction
  let draw: number | null = null;
  if (direction === "SHORT") {
    const below = ext.filter((p) => p.side === "sellside" && p.price < close);
    if (below.length) draw = Math.max(...below.map((p) => p.price));
  } else {
    const above = ext.filter((p) => p.side === "buyside" && p.price > close);
    if (above.length) draw = Math.min(...above.map((p) => p.price));
  }
  if (draw === null) return { ...story, reason: "no external draw on liquidity in storyline direction" };

  return {
    ...story,
    valid: true,
    reason: "ok",
    origin,
    drawOnLiquidity: draw,
    nearestExternalTarget: draw,
    internalPools: intp,
    externalPools: ext,
    imbalances: imb,
  };
}

/** Storyline snapshots at each recent closed H4 candle (point-in-time). */
export function storylineSeries(
  d1: Candle[], h4: Candle[], cfg: StrategyConfig, maxSnapshots = 90,
): [number, Storyline][] {
  const n = h4.length;
  const start = Math.max(20, n - maxSnapshots);
  const snaps: [number, Storyline][] = [];
  const day = 86400 * 1000;
  for (let j = start; j < n; j++) {
    const h4Close = h4[j].t + MAP_TF_SECONDS * 1000;
    const dSlice = d1.filter((d) => d.t + day <= h4Close);
    if (dSlice.length < 10) continue;
    snaps.push([h4[j].t, buildStoryline(dSlice, h4.slice(0, j + 1), cfg)]);
  }
  return snaps;
}

File: slk_bot/slk/storyline.py
"""Layer 1 — ABC storyline (expectation).

Built from the context timeframe (daily → weekly/monthly aggregates) and the
map timeframe (H4). A storyline is *valid* only when there is a directional
H4 environment, an opposing structure that has NOT been broken by a close,
an origin key level within reach of price, and a draw on liquidity to trade
towards.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from ..models import Candle, Direction
from . import features as F
from .types import Storyline

MAP_TF_SECONDS = 14400  # H4


def build_storyline(d1: list[Candle], h4: list[Candle], cfg) -> Storyline:
    """Build the storyline as-of the last candle of each input (callers slice
    inputs to avoid look-ahead when replaying)."""
    asof = h4[-1].time if h4 else datetime.now(timezone.utc)
    if len(h4) < 20:
        return Storyline(asof=asof, valid=False, reason="insufficient H4 data")
    if len(d1) < 10:
        return Storyline(asof=asof, valid=False, reason="insufficient daily data")

    env = F.environment(h4, cfg.pivot_left, cfg.pivot_right, cfg.min_swings_env)
    ph = F.phase(h4, env, cfg.phase_lookback, cfg.pivot_left, cfg.pivot_right)
    align = F.htf_alignment(d1, env, cfg.pivot_left, cfg.pivot_right)
    close = h4[-1].close

    story = Storyline(
        asof=asof, valid=False, environment=env, phase=ph,
        htf_alignment=align, map_close=close,
    )

    if env == "bullish":
        direction = Direction.LONG
    elif env == "bearish":
        direction = Direction.SHORT
    else:
        story.reason = "no directional H4 environment (consolidation)"
        return story
    story.direction = direction

    # HTF invalidation: a close through the opposing structure kills the story
    inv = F.structure_invalidation_level(h4, direction, cfg.pivot_left, cfg.pivot_right)
    if inv is not None:
        if direction is Direction.SHORT and close > inv:
            story.reason = f"H4 close {close:g} broke opposing structure {inv:g}"
            return story
        if direction is Direction.LONG and close < inv:
            story.reason = f"H4 close {close:g} broke opposing structure {inv:g}"
            return story

    atr_map = F.atr(h4, cfg.atr_period)
    ext = F.external_pools(d1, cfg.pivot_left, cfg.pivot_right)
    intp = F.internal_pools(h4, atr_map, cfg.decision_atr_mult,
                            cfg.pivot_left, cfg.pivot_right)
    imb = F.fvg_zones(h4, cfg.fvg_lookback)
    levels = F.key_levels(h4, cfg)
    F.mark_fvg_overlap(levels, imb)

    origin = F.select_origin(levels, close, atr_map, direction, cfg)
    if origin is None:
        story.reason = "no origin key level within reach of price"
        return story

    # draw on liquidity: nearest external pool in the environment direction
    if direction is Direction.SHORT:
        below = [p for p in ext if p.side == "sellside" and p.price < close]
        draw = max(below, key=lambda p: p.price) if below else None
    else:
        above = [p for p in ext if p.side == "buyside" and p.price > close]
        draw = min(above, key=lambda p: p.price) if above else None
    if draw is None:
        story.reason = "no external draw on liquidity in storyline direction"
        return story

    story.valid = True
    story.reason = "ok"
    story.origin = origin
    story.draw_on_liquidity = draw.price
    story.nearest_external_target = draw.price
    story.internal_pools = intp
    story.external_pools = ext
    story.imbalances = imb
    return story


def storyline_series(
    d1: list[Candle],
    h4: list[Candle],
    cfg,
    max_snapshots: int = 90,
) -> list[tuple[datetime, Storyline]]:
    """Storyline snapshots at each recent closed H4 candle, each built only
    from data closed at that moment (point-in-time correct for replay)."""
    n = len(h4)
    start = max(20, n - max_snapshots)
    snaps: list[tuple[datetime, Storyline]] = []
    day = timedelta(seconds=86400)
    for j in range(start, n):
        h4_close = h4[j].time + timedelta(seconds=MAP_TF_SECONDS)
        d_slice = [d for d in d1 if d.time + day <= h4_close]
        if len(d_slice) < 10:
            continue
        story = build_storyline(d_slice, h4[: j + 1], cfg)
        snaps.append((h4[j].time, story))
    return snaps
